---
title: 'Crash Course in Open Targets Part 3: Taming the Data Downloads'
author: CE West
date: '2021-07-29'
slug: crash-course-in-open-targets-part-3
categories:
  - science
  - python
tags:
  - bioinformatics
  - genetics
meta_img: images/image.png
cardtype: summary
description: An overview of accessing Open Targets data through the data downloads
draft: true
---

<script src="{{< blogdown/postref >}}index_files/header-attrs/header-attrs.js"></script>


<p>Now we’ve covered <a href="https://clarewest.github.io/blog/post/crash-course-in-open-targets-part-1/">the basics of Open Targets</a> and <a href="https://clarewest.github.io/blog/post/crash-course-in-open-targets-part-2/">dug deeper into the evidence</a> that makes up the data, this final part of the crash course covers how to harness it at a bigger scale through the data downloads.</p>
<!--more-->
<p>Now we’ve covered <a href="https://clarewest.github.io/blog/post/crash-course-in-open-targets-part-1/">the basics of Open Targets</a> and <a href="https://clarewest.github.io/blog/post/crash-course-in-open-targets-part-2/">dug deeper into the evidence</a> that makes up the data, this final part of the crash course covers how to harness it at a bigger scale through the data downloads.</p>
<p>For one whole year I happily hammered away at the APIs, but my queries kept getting bigger and more complex and eventually I wanted to use just about every piece of data that Open Targets provides. The Open Targets team finally told me (to paraphrase): “Clare, it’s time to grow up, the APIs aren’t designed for what you’re doing, use the data downloads instead”.</p>
<p>At first I was hesitant about handling big confusing files on my little laptop, but they were right - it was a lot easier and it only took me one day to figure it out. I’ve had no trouble querying these large datasets on my 8GB RAM laptop. Sadly, I have had to cancel my plans to crowdsource a new MacBook Pro.</p>
<div id="data-downloads" class="section level2">
<h2>Data Downloads</h2>
<p>I really recommend using using the <a href="https://platform.opentargets.org/downloads">data downloads</a> for Open Targets Platform if you are querying more than a few targets or diseases, especially if you are interested in lots of the annotations. While the Platform is extremely comprehensive, the data itself is not terribly big (&lt;6GB) and you can easily download the whole thing.</p>
<p>The downloads are available <a href="http://ftp.ebi.ac.uk/pub/databases/opentargets/platform/">via FTP</a> in JSON and parquet format, and you can find download instructions and usage examples in Python and R <a href="https://platform-docs.opentargets.org/data-access/datasets">here</a>. For more examples of how to use the data, check out <a href="https://community.opentargets.org/">Open Targets Community</a>.</p>
<p>For the Genetics Portal, the data is <a href="https://genetics-docs.opentargets.org/technical-pipeline/data-download">available to download</a> in JSON format <a href="https://ftp.ebi.ac.uk/pub/databases/opentargets/genetics/">via FTP</a>. It’s a lot bigger than the Platform - there are a few million variants after all - but you can always download specific parts of the data that you’re interested in processing yourself, and supplement it with API calls. For example, locally I have the directories:</p>
<ul>
<li><code>lut</code>: look-up tables, containing details on genes, studies, variants, and study-study overlaps</li>
<li><code>v2d</code>: <a href="https://genetics-docs.opentargets.org/our-approach/assigning-traits-to-loci">variant-to-disease</a> information i.e. GWAS study, variant, effect size, p-value, the trait for the study and the EFO disease codes it was mapped to, etc</li>
<li><code>v2d_coloc</code>: results of the <a href="https://genetics-docs.opentargets.org/our-approach/colocalisation-analysis">colocalisation analysis</a> performed between GWAS and eQTL/pQTL studies, as well as between pairs of GWAS studies with overlapping hits</li>
</ul>
<p>(that’s 56GB of my laptop dedicated to the Genetics Portal, but I am a superfan).</p>
</div>
<div id="fast-data-access-and-querying-using-spark" class="section level2">
<h2>Fast data access and querying using Spark</h2>
<p>Apache Spark is a superfast way to read and interact with the data downloads, and there are Spark interfaces available for Python (<a href="https://spark.apache.org/docs/latest/api/python/index.html">PySpark</a>) and R (<a href="https://spark.rstudio.com/">sparklyr</a>). If you’re used to functional programming, for example using the <code>tidyverse</code>, <code>pandas</code>, or querying SQL databases, then handling data using Spark should feel relatively familiar.</p>
<p>First, if you don’t already have it, you have to <a href="https://spark.apache.org/docs/latest/api/python/getting_started/index.html">install Spark and PySpark</a> (this is the hard part).</p>
<div id="reading-in-the-data" class="section level3">
<h3>Reading in the data</h3>
<p>Here’s an example of reading in Open Targets Platform data in Python using PySpark:</p>
<pre class="python"><code>import pyspark
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, explode

## Set up Spark
sc = pyspark.SparkContext()
spark = SparkSession.builder \
           .getOrCreate()

## Data paths
ot_platform = &quot;path/to/platform/files/21.02/&quot;
ot_genetics = &quot;path/to/portal/files/20022712/&quot;

## Read Open Targets platform data 
diseases = (spark.read.parquet(ot_platform+&quot;ETL_parquet/diseases/&quot;, header=True)
            .withColumnRenamed(&quot;id&quot;,&quot;diseaseId&quot;)
            .withColumnRenamed(&quot;name&quot;,&quot;diseaseName&quot;)
            )
targets = (spark.read.parquet(ot_platform+&quot;ETL_parquet/targets/&quot;)
           .withColumnRenamed(&quot;id&quot;,&quot;targetId&quot;)
           .withColumnRenamed(&quot;approvedSymbol&quot;, &quot;targetSymbol&quot;)
           .withColumnRenamed(&quot;approvedName&quot;,&quot;targetName&quot;)
           )
evidences = spark.read.parquet(ot_platform+&quot;ETL_parquet/evidences/succeeded&quot;)
knowndrugs = spark.read.parquet(ot_platform+&quot;ETL_parquet/knownDrugs&quot;)
overall_associations = spark.read.parquet(ot_platform+&quot;ETL_parquet/associations/indirect/byOverall/&quot;)

## Read Open Targets Genetics Portal data
studies = spark.read.json(ot_genetics+&quot;lut/study-index/&quot;)
v2d = spark.read.json(ot_genetics+&quot;v2d/&quot;)
coloc = spark.read.json(ot_genetics+&quot;v2d_coloc/&quot;)</code></pre>
<p>(Note that I rename some columns to make it easier to join them together later)</p>
</div>
<div id="inspecting-available-fields" class="section level3">
<h3>Inspecting available fields</h3>
<p>You can use <code>.printSchema()</code> to explore the data available for each dataframe. Let’s have a look at the fields for known drugs:</p>
<pre class="python"><code>knowndrugs.printSchema()
root
 |-- drugId: string (nullable = true)
 |-- targetId: string (nullable = true)
 |-- diseaseId: string (nullable = true)
 |-- phase: long (nullable = true)
 |-- status: string (nullable = true)
 |-- urls: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- niceName: string (nullable = true)
 |    |    |-- url: string (nullable = true)
 |-- ancestors: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- label: string (nullable = true)
 |-- approvedSymbol: string (nullable = true)
 |-- approvedName: string (nullable = true)
 |-- targetClass: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- prefName: string (nullable = true)
 |-- tradeNames: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- synonyms: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- drugType: string (nullable = true)
 |-- mechanismOfAction: string (nullable = true)
 |-- references: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- source: string (nullable = true)
 |    |    |-- ids: array (nullable = true)
 |    |    |    |-- element: string (containsNull = true)
 |    |    |-- urls: array (nullable = true)
 |    |    |    |-- element: string (containsNull = true)
 |-- targetName: string (nullable = true)</code></pre>
</div>
<div id="filtering-associations-and-evidence" class="section level3">
<h3>Filtering associations and evidence</h3>
<p>You can filter the data to get subsets that you’re interested in.</p>
<p>For example, if you’re only interested in target-disease <strong>associations</strong> that are based on genetic evidence, you can just read in associations from the datatype <code>genetic_association</code>:</p>
<pre class="python"><code>## Associations by data type (e.g. genetic association)
associations = spark.read.parquet(ot_platform+&quot;ETL_parquet/associations/indirect/byDatatype/&quot;)
gen_associations = associations.filter(col(&quot;datatypeId&quot;)==&quot;genetic_association&quot;)

gen_associations.printSchema()
root
 |-- targetId: string (nullable = true)
 |-- diseaseId: string (nullable = true)
 |-- datatypeId: string (nullable = true)
 |-- datatypeHarmonicScore: double (nullable = true)
 |-- datatypeEvidenceCount: long (nullable = true)
 |-- diseaseLabel: string (nullable = true)
 |-- targetName: string (nullable = true)
 |-- targetSymbol: string (nullable = true)</code></pre>
<p>You can also look at the <strong>evidence</strong> that supports associations, and filter based on a particular data source of interest. In this example, this will read in only evidence from the Genetics Portal (GWAS data):</p>
<pre class="python"><code># Evidence from one source (e.g. Open Targets Genetics evidence)
otg_evidences = spark.read.parquet(ot_platform+&quot;ETL_parquet/evidences/succeeded/sourceId\=ot_genetics_portal/&quot;)</code></pre>
</div>
<div id="getting-associations-and-evidence-for-targets-and-diseases" class="section level3">
<h3>Getting associations and evidence for targets and diseases</h3>
<p>You can easily subset and combine data by joining tables together through an ID column.</p>
<p>For example, say I have a dataframe containing my diseases of interest (in this case, Parkinson’s disease and Alzheimer’s disease):</p>
<pre class="python"><code>## My diseases of interest
my_diseases = spark.createDataFrame(
    [(&quot;EFO_0002508&quot;, &quot;My Disease ID 1&quot;), (&quot;EFO_0000249&quot;, &quot;My Disease ID 2&quot;)],
    [&quot;diseaseId&quot;,&quot;myDiseaseId&quot;]  
)

my_diseases.show()
+-----------+---------------+
|  diseaseId|    myDiseaseId|
+-----------+---------------+
|EFO_0002508|My Disease ID 1|
|EFO_0000249|My Disease ID 2|
+-----------+---------------+</code></pre>
<p>I can then join this with the <code>diseases</code> dataframe on the <code>diseaseId</code> column to get disease annotations from Open Targets for those diseases:</p>
<pre class="python"><code>my_diseases_details = (my_diseases.join(diseases, &quot;diseaseId&quot;, &quot;left&quot;)
                       .select(&quot;diseaseId&quot;, &quot;myDiseaseId&quot;, &quot;diseaseName&quot;, &quot;description&quot;, &quot;therapeuticAreas&quot;, &quot;synonyms&quot; )
                       )
            
my_diseases_details.printSchema()
root
 |-- diseaseId: string (nullable = true)
 |-- myDiseaseId: string (nullable = true)
 |-- diseaseName: string (nullable = true)
 |-- description: string (nullable = true)
 |-- therapeuticAreas: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- synonyms: struct (nullable = true)
 |    |-- hasBroadSynonym: array (nullable = true)
 |    |    |-- element: string (containsNull = true)
 |    |-- hasExactSynonym: array (nullable = true)
 |    |    |-- element: string (containsNull = true)
 |    |-- hasNarrowSynonym: array (nullable = true)
 |    |    |-- element: string (containsNull = true)
 |    |-- hasRelatedSynonym: array (nullable = true)
 |    |    |-- element: string (containsNull = true)</code></pre>
<p>To get the targets genetically associated with my diseases of interest, I join this with the <code>gen_associations</code> table, again on the <code>diseaseName</code> column. This will add all the columns from the genetic associations dataframe for the associations that involve my diseases of interest.</p>
<p>I can see how many targets are genetically associated with each disease by using <code>groupBy("diseaseName")</code> to group the rows by disease <code>count()</code> to count the number of rows in each group:</p>
<pre class="python"><code>my_diseases_associations = my_diseases_details.join(gen_associations,&quot;diseaseId&quot;, &quot;inner&quot;)

my_diseases_associations.groupBy(&quot;diseaseName&quot;).count().show()
+-------------------+-----+
|        diseaseName|count|
+-------------------+-----+
|Parkinson&#39;s disease|  327|
|Alzheimer&#39;s disease|  432|
+-------------------+-----+</code></pre>
<p>I can get the target annotations for those targets by joining with the <code>targets</code> dataframe on the <code>targetId</code>, <code>targetSymbol</code> and <code>targetName</code> columns, which are common to both dataframes.</p>
<pre class="python"><code>my_targets = my_diseases_associations.join(targets, [&quot;targetId&quot;, &quot;targetSymbol&quot;, &quot;targetName&quot;])</code></pre>
<p>I then have all the Open Targets annotations to explore for these targets. I won’t show them all (because there are so many!) but to demonstrate, here’s how you can access information on the <a href="https://docs.targetvalidation.org/getting-started/target-tractability">tractability category</a> and the functions of the targets:</p>
<pre class="python"><code>my_targets_details = (my_targets
                      .orderBy(col(&quot;datatypeHarmonicScore&quot;), 
                               ascending = False)
                      .select(&quot;targetSymbol&quot;, &quot;tractability.smallmolecule.top_category&quot;, 
                              explode(&quot;proteinAnnotations.functions&quot;))
                      )

my_targets_details.show(5, truncate = 40)
+------------+---------------------------------------+----------------------------------------+
|targetSymbol|tractability.smallmolecule.top_category|                                     col|
+------------+---------------------------------------+----------------------------------------+
|       PSEN1|                 Clinical_Precedence_sm|Catalytic subunit of the gamma-secret...|
|        PRKN|                Discovery_Precedence_sm|Functions within a multiprotein E3 ub...|
|       PINK1|                 Predicted_Tractable_sm|Protects against mitochondrial dysfun...|
|     ATP13A2|                                   null|ATPase which acts as a lysosomal poly...|
|       LRRK2|                Discovery_Precedence_sm|Serine/threonine-protein kinase which...|
+------------+---------------------------------------+----------------------------------------+</code></pre>
<p>Finally, let’s bring in some Genetics Portal data.</p>
<p>In the following example, I get GWAS evidence that links targets to my diseases of interest, where the locus-to-gene score is greater than 0.5. Then I use the <code>studyId</code> field to join it with the <code>studies</code> dataframe, which contains information on the GWAS studies from the Genetics Portal.</p>
<pre class="python"><code>my_diseases_evidence  = (my_diseases_associations
                    # join on the target and disease fields common to the association and evidence dataframes
                    .join(otg_evidences, [&quot;diseaseId&quot;, &quot;targetId&quot;, &quot;targetName&quot;, &quot;targetSymbol&quot;], &quot;inner&quot;)
                    .select(&quot;targetId&quot;, &quot;targetSymbol&quot;, &quot;diseaseId&quot;, &quot;diseaseName&quot;, &quot;studyId&quot;, &quot;variantId&quot;, &quot;score&quot;)
                    .filter(col(&quot;score&quot;)&gt;0.5)
                    )

my_diseases_studies = (my_diseases_evidence
                        .join(studies
                              .withColumnRenamed(&quot;study_id&quot;, &quot;studyId&quot;),
                              &quot;studyId&quot;)
                        )</code></pre>
</div>
<div id="conclusion" class="section level3">
<h3>Conclusion</h3>
<p>I hope you enjoyed this whistlestop tour of Open Targets. If you want to graduate to the next level, check out <a href="https://platform-docs.opentargets.org/data-access/google-bigquery">Google BigQuery access</a>. If you want to learn more about the data and the platforms, check out the documentation (<a href="https://platform-docs.opentargets.org/getting-started">Platform</a>; <a href="https://genetics-docs.opentargets.org/">Genetics Portal</a>) and publications [1, 2]. You can learn more about the <a href="https://platform-docs.opentargets.org/infrastructure">infrastructure</a> and <a href="https://github.com/opentargets">read the source code on GitHub</a>. And if you find any great drug targets, let me know!</p>
<ol style="list-style-type: decimal">
<li><p>Ochoa, D. et al. (2021). <a href="http://dx.doi.org/10.1093/nar/gkaa1027">Open Targets Platform: supporting systematic drug–target identification and prioritisation</a>. Nucleic Acids Research.</p></li>
<li><p>Ghoussaini, M et al. (2021). <a href="http://dx.doi.org/10.1093/nar/gkaa840">Open Targets Genetics: systematic identification of trait-associated genes using large-scale genetics and functional genomics</a>. Nucleic Acids Research.</p></li>
</ol>
</div>
</div>
